internal/backend/runtime/omni/audit/log_file.go

```diff
@@ -0,0 +1,87 @@
+// Copyright (c) 2024 Sidero Labs, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+
+package audit
+
+import (
+	"encoding/json"
+	"os"
+	"path/filepath"
+	"runtime"
+	"sync"
+	"time"
+
+	"github.com/siderolabs/gen/pair/ordered"
+)
+
+// LogFile is a rotating log file.
+//
+//nolint:govet
+type LogFile struct {
+	dir string
+
+	mu        sync.Mutex
+	f         *os.File
+	lastWrite time.Time
+}
+
+// NewLogFile creates a new rotating log file.
+func NewLogFile(dir string) *LogFile {
+	return &LogFile{dir: dir}
+}
+
+// Write writes data to the log file, creating new one on demand.
+func (l *LogFile) Write(data any) error {
+	return l.writeAt(data, time.Now())
+}
+
+func (l *LogFile) writeAt(data any, at time.Time) error {
+	f, err := l.openFile(at)
+	if err != nil {
+		return err
+	}
+
+	if err := json.NewEncoder(f).Encode(data); err != nil {
+		return err
+	}
+
+	l.mu.Lock()
+	l.lastWrite = at
+	l.mu.Unlock()
+
+	return nil
+}
+
+// openFile opens a file for the given date. It returns the file is date for at matches
+// the last write date. Otherwise, it opens a new file.
+func (l *LogFile) openFile(at time.Time) (*os.File, error) {
+	l.mu.Lock()
+	defer l.mu.Unlock()
+
+	// TODO(Dmitriy): decide what to do with when we write entry for the previous day
+	// into the current day (happens on day's border). One solution is to have two decriptors opened:
+	// one for the current day and one for the previous day. We can close the previous day's descriptor
+	// when we open the 3rd day's descriptor. This way we will have at most two descriptors opened.
+	if l.f != nil && ordered.MakeTriple(at.Date()).Compare(ordered.MakeTriple(l.lastWrite.Date())) <= 0 {
```

Comment:
idk, but it feels it might be easier to run a goroutine which "flips" the file at midnight, and all writes go to a single currently opn file, so they are at least chronologically ordered?

Reply:
I reworked this entirely, more explanation below.